libgalois/src/analytics/louvain_clustering/louvain_clustering.cpp

```diff
@@ -173,6 +173,200 @@ struct LouvainClusteringImplementation
+                    n_data_degree_wt, n_data_curr_comm_id,
+                    constant_for_second_term);
+
+              } else {
+                local_target[n] = Base::UNASSIGNED;
+              }
+
+              /* Update cluster info */
+              if (local_target[n] != n_data_curr_comm_id &&
+                  local_target[n] != Base::UNASSIGNED) {
+                katana::atomicAdd(
+                    c_update_add[local_target[n]].degree_wt, n_data_degree_wt);
+                katana::atomicAdd(
+                    c_update_add[local_target[n]].size, (uint64_t)1);
+                katana::atomicAdd(
+                    c_update_subtract[n_data_curr_comm_id].degree_wt,
+                    n_data_degree_wt);
+                katana::atomicAdd(
+                    c_update_subtract[n_data_curr_comm_id].size, (uint64_t)1);
+
+                if (!in_bag[local_target[n]]) {
+                  to_process.push(local_target[n]);
+                  in_bag[local_target[n]] = true;
+                }
+
+                if (!in_bag[n_data_curr_comm_id]) {
+                  to_process.push(n_data_curr_comm_id);
+                  in_bag[n_data_curr_comm_id] = true;
+                }
+              }
+            },
+            katana::loopname("louvain algo: Phase 1"));
+
+        katana::do_all(katana::iterate(bag[idx]), [&](GNode n) {
+          graph.template GetData<CurrentCommunityId>(n) = local_target[n];
+        });
+
+        for (auto n : to_process) {
+          if (in_bag[n]) {
+            katana::atomicAdd(c_info[n].size, c_update_add[n].size.load());
+            katana::atomicAdd(
+                c_info[n].degree_wt, c_update_add[n].degree_wt.load());
+
+            katana::atomicSub(c_info[n].size, c_update_subtract[n].size.load());
+            katana::atomicSub(
+                c_info[n].degree_wt, c_update_subtract[n].degree_wt.load());
+            c_update_add[n].size = 0;
+            c_update_add[n].degree_wt = 0;
+            c_update_subtract[n].size = 0;
+            c_update_subtract[n].degree_wt = 0;
+            in_bag[n] = false;
+          }
+        }
+
+      }  // end for
+
+      /* Calculate the overall modularity */
+      double e_xx = 0;
+      double a2_x = 0;
+
+      curr_mod = Base::template CalModularity<EdgeWeightType>(
+          graph, c_info, e_xx, a2_x, constant_for_second_term);
+
+      if ((curr_mod - prev_mod) < modularity_threshold_per_round) {
+        prev_mod = curr_mod;
+        break;
+      }
+
+      prev_mod = curr_mod;
+
+      if (prev_mod < lower)
+        prev_mod = lower;
+
+    }  // End while
+    TimerClusteringWhile.stop();
+
+    iter = num_iter;
+
+    c_info.destroy();
+    c_info.deallocate();
+
+    c_update_add.destroy();
+    c_update_add.deallocate();
+
+    c_update_subtract.destroy();
+    c_update_subtract.deallocate();
+
+    local_target.destroy();
+    local_target.deallocate();
+
+    in_bag.destroy();
+    in_bag.deallocate();
```

Comment:
`LargeArray` does all this in it's destructor. There is no need for this, here or in the original version. I wonder why this was written in the first place.